packet_encoder.go

```diff
@@ -22,6 +22,13 @@ type packetEncoder interface {
 	// Stacks, see PushEncoder
 	push(in pushEncoder)
 	pop() error
+
+	// Statistics in order to expose metrics
+	offset() int
+	recordTopic(topic string)
```

Comment:
I'm not sure why it's going through the packet-encoder at all? Can't the various callers (e.g. in `produce_request.go`) fetch and update the metric objects directly? They should have all the necessary context.